firebase/functions/index.js

```diff
@@ -27,118 +38,116 @@ exports.resultCounter = functions.database.ref('/v2/results/{projectId}/{groupId
-        .then((dataSnapshot) => {
-            const numberOfTasks = dataSnapshot.val()
-            return numberOfTasks
-        })
-        .then((numberOfTasks) => {
-            return userRef.child('taskContributionCount').transaction((currentCount) => {
-                return currentCount + numberOfTasks
-            })
-        })
-    promises.push(totalTaskContributionCount)
-
-    // Counter for User contributions on a distict Project
-    const groupContributionCount = userContributionRef.child('groupContributionCount').transaction((currentCount) => {
-        return currentCount + 1
-    });
-    promises.push(groupContributionCount)
-
-    const taskContributionCount = groupRef.child('numberOfTasks').once('value')
-        .then((dataSnapshot) => {
-            const numberOfTasks = dataSnapshot.val()
-            return numberOfTasks
-        })
-        .then((numberOfTasks) => {
-            return userContributionRef.child('taskContributionCount').transaction((currentCount) => {
-                return currentCount + numberOfTasks
-            })
-        })
-    promises.push(taskContributionCount)
-
-    const contributionTime = userContributionRef.child(context.params.groupId).once('value')
+    /*
+        Check if this user has submitted a results for this group already.
+        If no result has been submitted yet, set userId in v2/groupsUsers.
+        Then set this group contribution in the user profile.
+        Update overall taskContributionCount and project taskContributionCount in the user profile
+        based on the number of results submitted and the existing count values.
+    */
+    updateValues = groupUsersRef.child(context.params.userId).once('value')
         .then((dataSnapshot) => {
             if (dataSnapshot.exists()) {
                 console.log('group contribution exists already. user: '+context.params.userId+' project: '+context.params.projectId+' group: '+context.params.groupId)
                 return null
             }
             else {
+
                 const data = {
                     'startTime': result['startTime'],
                     'endTime': result['endTime']
                 }
-                return userContributionRef.child(context.params.groupId).set(data)
+                const numberOfTasks = Object.keys( result['results'] ).length
+
+                return {
+                    userContribution: userContributionRef.child(context.params.groupId).set(data),
+                    groupUsers: groupUsersRef.child(context.params.userId).set(true),
+                    totalTaskContributionCount: totalTaskContributionCountRef.transaction((currentCount) => {return currentCount + numberOfTasks}),
+                    taskContributionCount: taskContributionCountRef.transaction((currentCount) => {return currentCount + numberOfTasks})
+                }
             }
         })
-    promises.push(contributionTime)
+    promises.push(updateValues.userContribution)
```

Comment:
is this not going to crash if `updateValues` finds existing results and returns `null` (line 75/52)? This line here would try to read an attribute of `null`, which I'd expect to :boom:

Reply:
I updated this part of the code and check if `updateValues` returns null